### ISSUE TITLE
Decoding of coordinates into location name fails

### ISSUE THREAD

**Issue by tmp-hallenser** [LycheeOrg/Lychee-Docker#76]
The decoding of location coordinates into location name is failing (e.g. run `php artisan lychee:decode_GPS_locations`)

Log says:
```
2020-10-24 05:34:54 -- warning -- App\Metadata\Geodecoder::decodeLocation_core -- 84 -- Error creating resource: [message] fopen(): SSL operation failed with code 1. OpenSSL Error messages:
error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed
[file] /var/www/html/Lychee/vendor/guzzlehttp/guzzle/src/Handler/StreamHandler.php
[line] 324
[message] fopen(): Failed to enable crypto
[file] /var/www/html/Lychee/vendor/guzzlehttp/guzzle/src/Handler/StreamHandler.php
[line] 324
[message] fopen(https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=48.25700488&lon=10.08186616&addressdetails=1&zoom=18&accept-language=en): failed to open stream: operation failed
[file] /var/www/html/Lychee/vendor/guzzlehttp/guzzle/src/Handler/StreamHandler.php
[line] 324
    2020-10-24 05:34:54 -- warning -- App\Metadata\Geodecoder::decodeLocation_core -- 83 -- Decoding of location failed!
```

This seems to be an issue with the docker container (certificates?)

**Comment by d7415** [comment]
It decodes fine for me on v4.0.7 with wikimedia (default provider) and openstreetmap.org (your selection). I'll try `:dev`...

**Comment by d7415** [comment]
Looks like provider is just for map tiles. But `:dev` also works with mongolia.jpeg from the test suite.

**Comment by tmp-hallenser** [comment]
Finally found some more time for debugging. 

The following php code snipped fails:
```
<?php
$homepage = file_get_contents('https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=49.10856967&lon=-120.15566859&addressdetails=1&zoom=18&accept-language=en');
echo $homepage;
```
with
```
PHP Warning:  file_get_contents(): SSL operation failed with code 1. OpenSSL Error messages:
error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed in /var/www/html/Lychee/test.php on line 2
PHP Warning:  file_get_contents(): Failed to enable crypto in /var/www/html/Lychee/test.php on line 2
PHP Warning:  file_get_contents(https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=49.10856967&lon=-120.15566859&addressdetails=1&zoom=18&accept-language=en): failed to open stream: operation failed in /var/www/html/Lychee/test.php on line 2
```

I'm running it the following docker-compose file on a raspberry pi (armv7)

```
#-------------------------------------------
#  Docker Compose
# @RobLandry
# Repo : https://github.com/LycheeOrg/Lychee-Laravel-Docker
#-------------------------------------------

version: '3'

services:
  lychee_db:
    container_name: lychee_db
    image: linuxserver/mariadb:latest
    environment:
      - MYSQL_ROOT_PASSWORD= MYSQL_ROOT_PASSWORD
      - MYSQL_DATABASE=lychee
      - MYSQL_USER= MYSQL_USER
      - MYSQL_PASSWORD= MYSQL_PASSWORD
    expose: 
      - 3306
    volumes:
      - mysql:/var/lib/mysql
    networks:
      - lychee
    restart: unless-stopped

  lychee:
    image: lycheeorg/lychee:dev
    container_name: lychee
    ports:
      - 80:80
    volumes:
      - /home/pi/docker-files/docker-volumes/lychee-laravel/import:/import
      - /home/pi/docker-files/docker-volumes/lychee-laravel/conf:/conf
      - /home/pi/docker-files/docker-volumes/lychee-laravel/uploads:/uploads
      - /home/pi/docker-files/docker-volumes/lychee-laravel/sym:/sym
    networks:
      - lychee
    environment:
      #- PUID=1000
      #- PGID=1000
      # PHP timezone e.g. PHP_TZ=America/New_York
      - PHP_TZ=UTC
      #- APP_NAME=Laravel
      #- APP_ENV=local
      #- APP_DEBUG=true
      #- APP_URL=http://localhost
      #- LOG_CHANNEL=stack
      - DB_CONNECTION=mysql
      - DB_HOST=lychee_db
      - DB_PORT=3306
      - DB_DATABASE=lychee
      - DB_USERNAME= MYSQL_USER
      - DB_PASSWORD= MYSQL_PASSWORD
      #- DB_DROP_CLEAR_TABLES_ON_ROLLBACK=false
      #- DB_OLD_LYCHEE_PREFIX=''
      #- BROADCAST_DRIVER=log
      #- CACHE_DRIVER=file
      #- SESSION_DRIVER=file
      #- SESSION_LIFETIME=120
      #- QUEUE_DRIVER=sync
      #- SECURITY_HEADER_HSTS_ENABLE=false
      #- REDIS_HOST=127.0.0.1
      #- REDIS_PASSWORD=null
      #- REDIS_PORT=6379
      #- MAIL_DRIVER=smtp
      #- MAIL_HOST=smtp.mailtrap.io
      #- MAIL_PORT=2525
      #- MAIL_USERNAME=null
      #- MAIL_PASSWORD=null
      #- MAIL_ENCRYPTION=null
      #- PUSHER_APP_ID=''
      #- PUSHER_APP_KEY=''
      #- PUSHER_APP_SECRET=''
      #- PUSHER_APP_CLUSTER=mt1
      - STARTUP_DELAY=30
    restart: unless-stopped
    depends_on:
      - lychee_db

networks:
  lychee:

volumes:
  mysql:
```

**Comment by d7415** [comment]
Nothing's leaping out at me there. The fact that I can't reproduce makes me think it may be an ARM problem. Seems odd that the ARM build would have different certs though.

A couple of threads suggest that the same error arises if there's no connectivity at all (in their case outbound SMTP being eaten by antivirus). If you shell in can you connect to other websites with e.g. curl?

**Comment by tmp-hallenser** [comment]
The container has an internet connection. Ping for works. curl also work for example for www.google.com, but not for https.

```
curl 'https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=49.30856967&lon=-123.15566859&addressdetails=1&zoom=18&accept-language=en'
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.

```

**Comment by d7415** [comment]
Ok, if it's not environment, it makes it sound like an arm problem. Have you tried with `:v4.0.8` or `dev`? They should have both picked up the latest buster-slim last night.

**Comment by tmp-hallenser** [comment]
no change, still not working.

I've run manually the steps from the Dockerfile file on a blank image - work perfectly fine. Very strange.

**Comment by d7415** [comment]
Very odd...

**Comment by tmp-hallenser** [comment]
Finally found a solution. `update-ca-certificates -f -v` does the trick.

**Comment by d7415** [comment]
Even weirder, but good find. Does it work in the Dockerfile instead of entrypoint? I've opened #80 which is done that way instead, so there should be a `:testing` image with that soon(ish). Though that will be replaced by the cron run at some point.

**Comment by tmp-hallenser** [comment]
Just switched to testing branch and #80 is not working (very strange). Is there a way to test the docker image #79 (created PR to give it a try, but #80 has overwritten the dockerhub image before I had a chance)

**Comment by d7415** [comment]
It was probably overwritten by the cron. I'll try to force Travis to push to some separate tags for testing.

I was considering getting it to do that properly for all PRs a while ago, but never got around to doing it. Partially because it could only work for PRs from LycheeOrg).

**Comment by d7415** [comment]
This may be delayed due to Travis...

**Comment by tmp-hallenser** [comment]
> This may be delayed due to Travis...

Still not finished ...

**Comment by ildyria** [comment]
Well... @d7415 can try to use the github actions. :smile:

**Comment by d7415** [comment]
> Still not finished ...

I didn't say *how* delayed.

tl;dr Travis [changed](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing) their billing system but didn't announce it very well. We ran out of "credits" on Monday morning, took a few hours to figure out what was happening (and that they weren't going to replenish automatically, as implied by the website) and applied for some "OSS-only credits". That's apparently pending direction from management.

**Comment by d7415** [comment]
> Well... @d7415 can try to use the github actions. smile

Personally I'm all for that, but it would be bad for @tmp-hallenser as it wouldn't test arm :P

**Comment by d7415** [comment]
Ok, set up a quick, manual GitHub Action that should be building some images...

**Comment by d7415** [comment]
Right. @tmp-hallenser I have 3 images. Firstly, can you see if `lycheeorg/lychee:gha-bullseye` fixes it? If not, `lycheeorg/lychee:gha-pr80` is my attempt and `lycheeorg/lychee:gha-pr79` is yours

**Comment by tmp-hallenser** [comment]
`lycheeorg/lychee:gha-bullseye` is working, but I had to add `privileged: true` in the docker-compose file to mitigate the `sleep: cannot read realtime clock: Operation not permitted`

**Comment by d7415** [comment]
Ok. What version of Docker are you using. Supposedly the fix for that was added in 19.03.9.

**Comment by d7415** [comment]
In the meantime, I'm going to merge the bullseye branch and close the others. I haven't set GitHub Actions to push properly yet. I might if the Travis issues continue much longer.

**Comment by tmp-hallenser** [comment]
> Ok. What version of Docker are you using? Supposedly the fix for that was added in 19.03.9.

I'm on `Docker version 19.03.13, build 4484c46`

Spend some more time debugging (sorry for the too fast reply above). The `curl` command is now working on `bullseye` if I add `privileged: true`.

But, in order to get rid of the errors from post 1 for decoding the GPS location, I still need `update-ca-certificates -f -v`. (which I don't understand ...)

**Comment by d7415** [comment]
Ok, that's even weirder...

**Comment by tmp-hallenser** [comment]
`lycheeorg/lychee:gha-pr79` is working *yeah*

**Comment by tmp-hallenser** [comment]
#84 should do the job to fix this one.

**Comment by d7415** [comment]
Sorry. Forgot about this. Could you try #85? I'm still reluctant to put this in `entrypoint.sh` unless we have to as it will slow down every boot of the image on every architecture.

**Comment by tmp-hallenser** [comment]
quick feedback:

lychee:testing_1122 and testing_1123 have the same symptoms (hope one of them was the right one):
* curl command is working
* php file_get_contents not

If you're worried about start-up times, what about adding 

```
// check if we're on a armv7 architecture - docker image behaves strange and we need to update certificates
if [[ $(uname -a) == *"armv7"* ]]; then
  update-ca-certificates -f -v
fi
```

**Comment by d7415** [comment]
Yeah, looks like it was 1122.

I like the `if` idea. I'll try to sort that over the next few days and _then_ push 4.1.0 out.
